File: FortunaWeb/fortuna/src/main/java/it/unibo/ingsoft/fortuna/model/Sconto.java
package it.unibo.ingsoft.fortuna.model;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

public class Sconto {
    private LocalDateTime inizio;
    private LocalDateTime fine;
    private double quantita;
    private double quantitaPct;
    private double costoMinimo;

    private Set<Prodotto> perProdotti;

    public Sconto() {}

    public static Sconto of(LocalDateTime inizio, LocalDateTime fine, double quantita, double quantitaPct, double costoMinimo) {
        Sconto sconto = new Sconto();
        return sconto.inizio(inizio)
            .fine(fine)
            .quantita(quantita)
            .quantitaPct(quantitaPct)
            .costoMinimo(costoMinimo);
    }

    public static Sconto ofProdotti(LocalDateTime inizio, LocalDateTime fine, double quantita, double quantitaPct, double costoMinimo, Set<Prodotto> perProdotti) {
        return of(inizio, fine, quantita, quantitaPct, costoMinimo)
            .perProdotti(perProdotti);
    }

    public static Sconto ofProdotti(LocalDateTime inizio, LocalDateTime fine, double quantita, double quantitaPct, double costoMinimo, Prodotto perProdotto) {
        Set<Prodotto> prodotti = new HashSet<Prodotto>();
        prodotti.add(perProdotto);
        return of(inizio, fine, quantita, quantitaPct, costoMinimo)
            .perProdotti(prodotti);
    }

    public boolean isAttivo(LocalDateTime tempo, Prodotto prodotto, double costoTotale)
    {
        return isDateCostGood(tempo, costoTotale) && (perProdotti != null && perProdotti.contains(prodotto));
    }

    public boolean isAttivo(LocalDateTime tempo, double costoTotale)
    {
        return isDateCostGood(tempo, costoTotale) && (perProdotti == null || perProdotti.isEmpty());
    }

    private boolean isDateCostGood(LocalDateTime tempo, double costoTotale)
    {
        return (tempo.isAfter(inizio) || tempo.isEqual(inizio)) && tempo.isBefore(fine) && costoTotale >= costoMinimo;
    }



    public LocalDateTime getInizio() {
        return this.inizio;
    }

    public void setInizio(LocalDateTime inizio) {
        this.inizio = inizio;
    }

    public LocalDateTime getFine() {
        return this.fine;
    }

    public void setFine(LocalDateTime fine) {
        this.fine = fine;
    }

    public double getQuantita() {
        return this.quantita;
    }

    public void setQuantita(double quantita) {
        this.quantita = quantita;
    }

    public double getQuantitaPct() {
        return this.quantitaPct;
    }

    public void setQuantitaPct(double quantitaPct) {
        this.quantitaPct = quantitaPct;
    }

    public double getCostoMinimo() {
        return this.costoMinimo;
    }

    public void setCostoMinimo(double costoMinimo) {
        this.costoMinimo = costoMinimo;
    }

    public Set<Prodotto> getPerProdotti() {
        return this.perProdotti;
    }

    public void setPerProdotti(Set<Prodotto> perProdotti) {
        this.perProdotti = perProdotti;
    }


    public Sconto inizio(LocalDateTime inizio) {
        setInizio(inizio);
        return this;
    }

    public Sconto fine(LocalDateTime fine) {
        setFine(fine);
        return this;
    }

    public Sconto quantita(double quantita) {
        setQuantita(quantita);
        return this;
    }

    public Sconto quantitaPct(double quantitaPct) {
        setQuantitaPct(quantitaPct);
        return this;
    }

    public Sconto costoMinimo(double costoMinimo) {
        setCostoMinimo(costoMinimo);
        return this;
    }

    public Sconto perProdotti(Set<Prodotto> perProdotti) {
        setPerProdotti(perProdotti);
        return this;
    }
}

File: IngsoftC/RistoranteWeb/Models/Sconto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace RistoranteWeb.Models
{
    public class Sconto
    {
        private DateTime _inizio;
        private DateTime _fine;
        private double _quantita;
        private double _quantitaPct;
        private double _costoMinimo;

        private ISet<Prodotto> _perProdotti;

        //inserire costruttori e roba lol

        public DateTime Inizio { get => _inizio; set => _inizio = value; }
        public DateTime Fine { get => _fine; set => _fine = value; }
        public double Quantita { get => _quantita; set => _quantita = value; }
        public double QuantitaPct { get => _quantitaPct; set => _quantitaPct = value; }
        public double CostoMinimo { get => _costoMinimo; set => _costoMinimo = value; }
        public ISet<Prodotto> PerProdotti { get => _perProdotti; set => _perProdotti = value; }

        public bool IsAttivo(DateTime tempo, Prodotto prodotto, double costoTotale)
        {
            return isDateCostGood(tempo, costoTotale) && PerProdotti.Contains(prodotto);
        }

        public bool IsAttivo(DateTime tempo, double costoTotale)
        {
            return isDateCostGood(tempo, costoTotale) && PerProdotti.Count == 0;
        }

        private bool isDateCostGood(DateTime tempo, double costoTotale)
        {
            return tempo >= Inizio && tempo < Fine && costoTotale >= CostoMinimo;
        }
    }
}